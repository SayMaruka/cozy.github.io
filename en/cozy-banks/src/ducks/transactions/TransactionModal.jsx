/**
 * Is used in mobile/tablet mode when you click on the more button
 */

import React, { useState } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'

import {
  Icon,
  Spinner,
  Bd,
  Img,
  Alerter,
  useViewStack,
  ModalContent,
  useI18n,
  useBreakpoints,
  Caption,
  Chip
} from 'cozy-ui/transpiled/react'

import { withRouter, Link } from 'react-router'
import ModalStack from 'components/ModalStack'

import Figure from 'cozy-ui/transpiled/react/Figure'
import { PageModal } from 'components/PageModal'
import { PageHeader, PageBackButton } from 'components/PageModal/Page'

import CategoryIcon from 'ducks/categories/CategoryIcon'
import { getCategoryName } from 'ducks/categories/categoriesMap'
import { getCategoryId } from 'ducks/transactions/helpers'

import { getLabel } from 'ducks/transactions'
import TransactionActions from 'ducks/transactions/TransactionActions'
import { updateApplicationDate } from 'ducks/transactions/helpers'

import styles from 'ducks/transactions/TransactionModal.styl'
import { getCurrencySymbol } from 'utils/currencySymbol'

import iconCredit from 'assets/icons/icon-credit.svg'
import iconCalendar from 'assets/icons/icon-calendar.svg'
import iconRecurrence from 'assets/icons/icon-recurrence.svg'
import {
  getAccountLabel,
  getAccountInstitutionLabel
} from 'ducks/account/helpers'
import { TRANSACTION_DOCTYPE } from 'doctypes'
import flag from 'cozy-flags'
import { getDate, getApplicationDate } from 'ducks/transactions/helpers'

import TransactionCategoryEditor from './TransactionCategoryEditor'
import TransactionApplicationDateEditor from './TransactionApplicationDateEditor'
import TransactionRecurrenceEditor from 'ducks/transactions/TransactionRecurrenceEditor'

import { getFrequencyText } from 'ducks/recurrence/utils'
import TransactionModalRow, {
  TransactionModalRowIcon,
  TransactionModalRowMedia,
  RowArrow
} from './TransactionModalRow'

import withDocs from 'components/withDocs'

const TransactionLabel = ({ label }) => (
  <div className={styles.TransactionLabel}>{label}</div>
)

const TransactionInfo = ({ label, value }) => (
  <div className={styles.TransactionInfo}>
    <span className={styles.TransactionInfoLabel}>{label} :</span>
    {value}
  </div>
)

const TransactionInfos = ({ infos }) => (
  <div>
    {infos.map(({ label, value }) => (
      <TransactionInfo key={label} label={label} value={value} />
    ))}
  </div>
)

const withTransaction = withDocs(ownProps => ({
  transaction: [TRANSACTION_DOCTYPE, ownProps.transactionId]
}))

const TransactionCategoryEditorSlide = ({ transaction }) => {
  const { t } = useI18n()
  const { stackPop } = useViewStack()
  const { isMobile } = useBreakpoints()
  return (
    <>
      <PageHeader dismissAction={stackPop}>
        {isMobile ? <PageBackButton onClick={stackPop} /> : null}
        {t('Categories.choice.title')}
      </PageHeader>
      <ModalContent className="u-p-0">
        <TransactionCategoryEditor
          beforeUpdate={stackPop}
          onCancel={stackPop}
          transaction={transaction}
        />
      </ModalContent>
    </>
  )
}

const TransactionApplicationDateEditorSlide = ({
  transaction,
  beforeUpdate,
  afterUpdate
}) => {
  const { t } = useI18n()
  const { stackPop } = useViewStack()
  const handleBeforeUpdate = () => {
    beforeUpdate()
    stackPop()
  }

  return (
    <div>
      <PageHeader dismissAction={stackPop}>
        {t('Transactions.infos.chooseApplicationDate')}
      </PageHeader>
      <TransactionApplicationDateEditor
        beforeUpdate={handleBeforeUpdate}
        afterUpdate={afterUpdate}
        transaction={transaction}
      />
    </div>
  )
}

export const showAlertAfterApplicationDateUpdate = (transaction, t, f) => {
  const date = getApplicationDate(transaction) || getDate(transaction)
  Alerter.success(
    t('Transactions.infos.applicationDateChangedAlert', {
      applicationDate: f(date, 'MMMM')
    })
  )
}

const stopPropagation = ev => ev.stopPropagation()

const RecurrenceRow = withRouter(({ transaction, onClick, router }) => {
  const recurrence = transaction.recurrence && transaction.recurrence.data
  const { t } = useI18n()

  const recurrenceRoute = recurrence ? `/recurrence/${recurrence._id}` : null

  return (
    <TransactionModalRowMedia
      align={recurrence ? 'top' : undefined}
      onClick={onClick}
    >
      <Img>
        <TransactionModalRowIcon icon={iconRecurrence} />
      </Img>
      <Bd>
        <div>
          {recurrence
            ? t('Recurrence.choice.recurrent')
            : t('Recurrence.choice.not-recurrent')}
          {recurrence ? (
            <>
              <br />
              <Caption>{getFrequencyText(t, recurrence)}</Caption>
              {router.location.pathname !== recurrenceRoute ? (
                <Link to={recurrenceRoute}>
                  <Chip
                    onClick={stopPropagation}
                    variant="outlined"
                    size="small"
                    className="u-w-100 u-ph-2 u-mt-half u-flex-justify-center"
                  >
                    {t('Recurrence.see-transaction-history')}
                  </Chip>
                </Link>
              ) : null}
            </>
          ) : null}
        </div>
      </Bd>
      <Img>
        <RowArrow />
      </Img>
    </TransactionModalRowMedia>
  )
})

const TransactionRecurrenceEditorSlide = ({ transaction }) => {
  const { t } = useI18n()

  const { stackPop } = useViewStack()

  return (
    <div>
      <PageHeader dismissAction={stackPop}>
        {t('Transactions.infos.chooseRecurrence')}
      </PageHeader>
      <ModalContent className="u-p-0">
        <TransactionRecurrenceEditor
          onSelect={x => x}
          beforeUpdate={stackPop}
          onCancel={stackPop}
          transaction={transaction}
        />
      </ModalContent>
    </div>
  )
}

/**
 * Show information of the transaction
 */
const TransactionModalInfoContent = withTransaction(props => {
  const { t, f } = useI18n()
  const { stackPush } = useViewStack()
  const { transaction, client, ...restProps } = props

  const typeIcon = (
    <Icon
      icon={iconCredit}
      width={16}
      className={cx({
        [styles['TransactionModalRowIcon-reversed']]: transaction.amount < 0
      })}
    />
  )

  const categoryId = getCategoryId(transaction)
  const account = transaction.account.data

  const handleShowCategoryChoice = () => {
    stackPush(<TransactionCategoryEditorSlide transaction={transaction} />)
  }

  const handleShowRecurrenceChoice = () => {
    stackPush(<TransactionRecurrenceEditorSlide transaction={transaction} />)
  }

  const [applicationDateBusy, setApplicationDateBusy] = useState(false)

  const handleAfterUpdateApplicationDate = updatedTransaction => {
    setApplicationDateBusy(false)
    showAlertAfterApplicationDateUpdate(updatedTransaction, t, f)
  }

  const handleShowApplicationEditor = ev => {
    !ev.defaultPrevented &&
      stackPush(
        <TransactionApplicationDateEditorSlide
          beforeUpdate={() => setApplicationDateBusy(true)}
          afterUpdate={handleAfterUpdateApplicationDate}
          transaction={transaction}
        />,
        { size: 'xsmall', mobileFullscreen: false }
      )
  }

  const handleResetApplicationDate = async ev => {
    ev.preventDefault()
    try {
      setApplicationDateBusy(true)
      const newTransaction = await updateApplicationDate(
        client,
        transaction,
        null
      )
      showAlertAfterApplicationDateUpdate(newTransaction, t, f)
    } finally {
      setApplicationDateBusy(false)
    }
  }

  const shouldShowRestoreApplicationDateIcon =
    getApplicationDate(transaction) && !applicationDateBusy

  return (
    <div className={styles['Separated']}>
      <TransactionModalRow iconLeft={typeIcon} align="top">
        <TransactionLabel label={getLabel(transaction)} />
        <TransactionInfos
          infos={[
            {
              label: t('Transactions.infos.account'),
              value: getAccountLabel(account)
            },
            {
              label: t('Transactions.infos.institution'),
              value: getAccountInstitutionLabel(account)
            },
            {
              label: t('Transactions.infos.originalBankLabel'),
              value: flag('originalBankLabel') && transaction.originalBankLabel
            },
            {
              label: t('Transactions.infos.date'),
              value: f(getDate(transaction), 'dddd D MMMM')
            }
          ].filter(x => x.value)}
        />
      </TransactionModalRow>
      <TransactionModalRowMedia onClick={handleShowCategoryChoice}>
        <Img>
          <CategoryIcon categoryId={categoryId} />
        </Img>
        <Bd>
          {t(
            `Data.subcategories.${getCategoryName(getCategoryId(transaction))}`
          )}
        </Bd>
        <Img>
          <RowArrow />
        </Img>
      </TransactionModalRowMedia>
      <TransactionModalRowMedia onClick={handleShowApplicationEditor}>
        <TransactionModalRowIcon icon={iconCalendar} />
        <Bd>
          {t('Transactions.infos.assignedToPeriod', {
            date: f(
              getApplicationDate(transaction) || getDate(transaction),
              'MMM YYYY'
            )
          })}
        </Bd>
        {shouldShowRestoreApplicationDateIcon ? (
          <Img>
            <span
              onClick={handleResetApplicationDate}
              className="u-expanded-click-area"
            >
              <Icon color="var(--slateGrey)" icon="restore" />
            </span>
          </Img>
        ) : null}
        {applicationDateBusy ? (
          <Img>
            <Spinner />
          </Img>
        ) : null}
        <Img>
          <RowArrow />
        </Img>
      </TransactionModalRowMedia>
      {flag('banks.recurrence') ? (
        <RecurrenceRow
          transaction={transaction}
          onClick={handleShowRecurrenceChoice}
        />
      ) : null}
      <TransactionActions
        transaction={transaction}
        {...restProps}
        displayDefaultAction
        isModalItem
      />
    </div>
  )
})

const TransactionModalInfoHeader = withTransaction(({ transaction }) => (
  <Figure
    total={transaction.amount}
    symbol={getCurrencySymbol(transaction.currency)}
    signed
  />
))

const TransactionModalInfo = props => {
  const { isMobile } = useBreakpoints()
  return (
    <div>
      <PageHeader dismissAction={props.requestClose}>
        {isMobile ? <PageBackButton onClick={props.requestClose} /> : null}
        <TransactionModalInfoHeader {...props} />
      </PageHeader>
      <TransactionModalInfoContent {...props} />
    </div>
  )
}

const TransactionModal = ({ requestClose, ...props }) => {
  const { t } = useI18n()
  return (
    <PageModal
      aria-label={t('Transactions.infos.modal-label')}
      dismissAction={requestClose}
      into="body"
      overflowHidden
    >
      <ModalStack>
        <TransactionModalInfo {...props} requestClose={requestClose} />
      </ModalStack>
    </PageModal>
  )
}

TransactionModal.propTypes = {
  requestClose: PropTypes.func.isRequired,
  transactionId: PropTypes.string.isRequired
}

export default TransactionModal
