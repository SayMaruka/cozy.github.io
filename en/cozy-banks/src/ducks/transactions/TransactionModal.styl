@require 'settings/breakpoints'

.Separated > * + *
    border-top 1px solid var(--silver)

.TransactionModalRow
    min-height 57px
    box-sizing border-box
    padding: 12px 24px

.TransactionModalRow-disabled
    color var(--coolGrey)
    font-style italic
    cursor default

.TransactionModalRowIcon-reversed
    transform rotate(180deg)

.TransactionLabel
    font-weight bold
    line-height 1.5
    color var(--charcoalGrey)

.TransactionInfo
    line-height 1.7
    font-size 14px
    color var(--coolGrey)

.TransactionInfoLabel
    display inline-block
    margin-right 8px
    color var(--charcoalGrey)
