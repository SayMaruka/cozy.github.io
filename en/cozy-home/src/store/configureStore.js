import { compose, createStore, applyMiddleware } from 'redux'
import { cozyMiddleware } from 'lib/redux-cozy-client'
import { createLogger } from 'redux-logger'
import konnectorsI18nMiddleware from 'lib/middlewares/konnectorsI18n'
import thunkMiddleware from 'redux-thunk'

import HomeStore from 'lib/HomeStore'
import flag from 'cozy-flags'
import getReducers from 'reducers'

const configureStore = (legacyClient, cozyClient, context, options = {}) => {
  // Enable Redux dev tools
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE || compose

  const reduxStore = createStore(
    getReducers(),
    composeEnhancers(
      applyMiddleware.apply(
        this,
        [
          cozyMiddleware(legacyClient),
          konnectorsI18nMiddleware(options.lang),
          thunkMiddleware,
          flag('redux-logger') ? createLogger() : null
        ].filter(Boolean)
      )
    )
  )

  return Object.assign(new HomeStore(context, cozyClient, options), reduxStore)
}

export default configureStore
