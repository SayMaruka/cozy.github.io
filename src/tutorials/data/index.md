#Â Introduction

This documentation is for developers that want to learn how to manipulate data in Cozy. Both front-end and back-end developers might find interest in this section: manipulating data cover a broad scope of concepts and traps to avoid.

This is organized as follows:

- [Queries](./queries.md)
- [PouchDB](./pouchdb.md) 
- [Doctypes](./doctypes.md)
- [Qualification](./qualification.md)
- [Advanced](./advanced.md)
