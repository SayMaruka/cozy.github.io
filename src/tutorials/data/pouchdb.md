# PouchDB - Offline

[PouchDB](https://pouchdb.com/) is a JavaScript implementation of CouchDB. It is particulary well suited for client environments such as browsers or mobile.

ℹ️ In Cozy, PouchDB is used in the desktop client and in mobile apps to synchronize data with the CouchDB server. Having a local database enables offline support and improves performances: the client can directly query local data to avoid network latency or unavailability.

While it is meant to mimic most of the CouchDB API, there are specificities that are good to know when working with PouchDB.

The findings below were obtained while working on sorting performance in Cozy Drive. For reference, the final PR [can be seen here](https://github.com/cozy/cozy-drive/pull/1002/files).

## Understanding what's going on

Before diving into some of the quirks, it's important to understand some things when it comes to Pouchdb and especially Mango queries.
First, you can add a plugin called `pouchdb-debug` and enable extra logs with `PouchDB.debug.enable( "pouchdb:find" );`. This will add explanation about the queries you run in the console and it's very helpful to understand what's going on under the hood.

You will realize that Pouchdb operates in 2 phases: first, it loads candidate documents in memory, and then it analyzes the candidates and only keeps the relevant ones. Long story short: the larger the number of candidates, the longer the operation will take. By default, all the documents in a database are candidates, but it is possible to have fewer candidates by using an index: by combining the query and the index, PouchDB can load only a subset of all the documents and may even skip the analyze phase if the index is really well suited to the query. A more detailed guide can be found [here](https://www.bennadel.com/blog/3258-understanding-the-query-plan-explained-by-the-find-plugin-in-pouchdb-6-2-0.htm).

## About indexes

Creating an index takes some time, but the first query will _also_ take time — you are encouraged to warm up the indexes by firing a query that uses it before it is actually needed. CozyPouchLink provides a way to automatically fire those queries after replication.

```js
import fromPairs from 'lodash/fromPairs'
import CozyClient, { Q } from 'cozy-client'
import CozyStackClient from 'cozy-stack-client'
import CozyPouchLink from 'cozy-pouch-link'

const TRANSACTION_DOCTYPE = 'io.cozy.bank.operations'

const makeWarmupQueryOptions = (doctype, indexedFields) => {
  return {
    definition: () => {
      const qdef = Q(doctype)
        .where(
          fromPairs(indexedFields.map(fieldName => [fieldName, { $gt: null }]))
        )
        .indexFields(indexedFields)
      return qdef
    },
    options: {
      as: `${doctype}-by-${indexedFields.join('-')}`
    }
  }
}

const pouchLink = new CozyPouchLink({
  doctypes: [TRANSACTION_DOCTYPE],
  doctypesReplicationOptions: {
    [TRANSACTION_DOCTYPE]: {
      warmupQueries: [
        makeWarmupQueryOptions(TRANSACTION_DOCTYPE, ['date']),
        makeWarmupQueryOptions(TRANSACTION_DOCTYPE, ['account']),
        makeWarmupQueryOptions(TRANSACTION_DOCTYPE, ['date', 'account'])
      ]
    }
  },
  initialSync: true
})

const client = new CozyClient({
  links: [pouchLink, new CozyStackClient()]
})
```

ℹ️ By default, Pouch will try to find the best index to use on your query. But for advanced queries, you generally want to force it with the use_index option to be sure it does not mess up.
ℹ️ If the query and the index you force are not compatible, Pouch will emit an error and not run the query at all.
ℹ️ If there is a change in the underlying documents, the index will be partially recalculated on the next query.

## Indexing more than one field

Creating an index on several fields is _not_ the same as creating multiple indexes on one field each. The effects of a single index on multiple fields is illustrated in the [official docs](https://pouchdb.com/guides/mango-queries.html#more-than-one-field) and is important to understand.
Furthermore, the order in which fields are indexed on a multi-index is significant, most notably when it comes to sorting. If you declare an index on the fields `['name', 'age']`, you should also sort them by `['name', 'age']`. Sorting them in a different order or using other fields will likely be done in memory and kill your performance.

See this [section](./advanced.md#indexes-performances-and-design) for more details and understand how to efficiently design indexes.

## Avoiding full scan

Filtering results with a selector on a field that has not been indexed is almost guaranteed to be done through a full scan and should be avoided. Since you can't have too many indexes, some filtering may have to be done in your own code — but if you can narrow down the results beforehand, that shouldn't be a problem.

Note that even selectors on indexed field may end up being done in memory: see this [section](./advanced.md#indexes-performances-and-design) for more details and understand how to avoid memory selectors.
