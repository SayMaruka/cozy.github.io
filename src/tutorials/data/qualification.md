# Document qualification

A document can be qualified by adding some attributes in its metadata meant to semantically describe the document. The objective is to maintain a transversal qualification structure between doctypes, so that files, banks, notes and so on could be qualified and retrieved. Currently, only the `io.cozy.files` doctype is supported. 

A qualification model is defined [here](https://github.com/cozy/cozy-client/blob/master/packages/cozy-client/src/assets/qualifications.json): it is used to provide a set of qualifications based on documents labels, so developers don't have to worry about defining their own qualification and to ensure qualifications consistency.

Note this is not an exhaustive nor closed model and [contributions](https://github.com/cozy/cozy-client/pulls) are welcomed.

To know more about the qualification format and a description of the qualification attributes with examples, see the [metadata documentation](https://github.com/cozy/cozy-doctypes/blob/master/docs/io.cozy.files_metadata.md).

## Qualify with cozy-client

This snippet shows how to qualify an health invoice with [cozy-client](https://github.com/cozy/cozy-client/):

```js
import { models } from 'cozy-client'
const { Qualification } = models.document
const { saveFileQualification } = models.file

const qualification = Qualification.getByLabel('health_invoice')
await saveFileQualification(client, file, qualification)
```

Then, then file metadata will include the following qualification:
```json
{
  "metadata": {
    "qualification": {
      "label": "health_invoice",
      "purpose": "invoice",
      "sourceCategory": "health"
    }
  }
}
```

Note `saveFileQualification` is a cozy-client helper that can be replaced by:

```js
const qualifiedFile = Document.setQualification(file, qualification)
await client
    .collection('io.cozy.files')
    .updateMetadataAttribute(file._id, qualifiedFile.metadata)

```
