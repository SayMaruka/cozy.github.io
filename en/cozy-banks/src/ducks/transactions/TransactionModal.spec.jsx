/* global mount */

import React from 'react'
import { DumbTransactionModal as TransactionModal } from './TransactionModal'
import data from '../../../test/fixtures'
import AppLike from 'test/AppLike'
import store, { normalizeData } from 'test/store'
import pretty from 'pretty'
import getClient from 'test/client'
import pick from 'lodash/pick'

const allTransactions = data['io.cozy.bank.operations']

describe('transaction modal', () => {
  let client
  beforeEach(() => {
    client = getClient()
    const documents = normalizeData(
      pick(data, 'io.cozy.bank.operations', 'io.cozy.bank.accounts')
    )
    jest
      .spyOn(client, 'getDocumentFromState')
      .mockImplementation((doctype, id) => {
        return documents[doctype][id]
      })
  })
  it('should render correctly', () => {
    // need to fix number of the account otherwise its randomly
    // set by the fixture
    const transaction = {
      ...allTransactions[0],
      _id: '2',
      _type: 'io.cozy.bank.operations'
    }

    const root = mount(
      <AppLike store={store} client={client}>
        <TransactionModal
          transactionId={transaction._id}
          transaction={client.hydrateDocument(transaction)}
          showCategoryChoice={() => {}}
          requestClose={() => {}}
          urls={{}}
          brands={[]}
        />
      </AppLike>
    )
    expect(pretty(root.html())).toMatchSnapshot()
  })
})
