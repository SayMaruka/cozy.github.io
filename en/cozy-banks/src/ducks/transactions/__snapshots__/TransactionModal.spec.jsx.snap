// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction modal should render correctly 1`] = `
"<div role=\\"dialog\\" class=\\"MuiModal-root-15 MuiDialog-root-1\\" aria-labelledby=\\"modal-title-0\\">
  <div class=\\"MuiBackdrop-root-17\\" aria-hidden=\\"true\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"></div>
  <div class=\\"MuiDialog-container-4 MuiDialog-scrollPaper-2\\" role=\\"document\\" style=\\"opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\" tabindex=\\"-1\\">
    <div class=\\"MuiPaper-root-19 MuiPaper-elevation24-45 MuiPaper-rounded-20 MuiDialog-paper-5 medium  MuiDialog-paperScrollPaper-6 MuiDialog-paperWidthSm-9\\"><button class=\\"styles__IconButton___1LWwV styles__DialogCloseButton___cxKPO\\"><svg viewBox=\\"0 0 16 16\\" class=\\"styles__icon___23x3R\\" width=\\"16\\" height=\\"16\\">
          <path fill-rule=\\"evenodd\\" d=\\"M8 6.59l5.3-5.3a1 1 0 011.4 1.42L9.42 8l5.3 5.3a1 1 0 01-1.42 1.4L8 9.42l-5.3 5.3a1 1 0 11-1.4-1.42L6.58 8l-5.3-5.3A1 1 0 112.7 1.3L8 6.58z\\"></path>
        </svg></button>
      <div class=\\"MuiDialogTitle-root-46 u-ellipsis\\" id=\\"modal-title-0\\">
        <div class=\\"u-ta-center\\">
          <div class=\\"\\"><span class=\\"Figure__Figure-total___MZ7Xt\\">-77.50</span><span class=\\"Figure__Figure-currency___195-B Figure__Figure__currency--withSpacing___2eXTy\\">â‚¬</span></div>
        </div>
      </div>
      <ul class=\\"MuiList-root-47 MuiList-padding-48\\">
        <li class=\\"MuiListItem-root-51 MuiListItem-default-54 MuiListItem-gutters-59 MuiListItem-alignItemsFlexStart-56\\">
          <div class=\\"MuiListItemIcon-root-64 WithStyles-ListItemIcon--root-63\\"><svg width=\\"16\\" height=\\"16\\" class=\\"TransactionModalRowIcon-reversed u-mt-1-half styles__icon___23x3R\\">
              <g fill=\\"#32363F\\" fill-rule=\\"evenodd\\">
                <path fill-rule=\\"nonzero\\" d=\\"M3.414 10H11a1 1 0 010 2H3.414l1.293 1.293a1 1 0 01-1.414 1.414l-3-3a1 1 0 010-1.414l3-3a1 1 0 011.414 1.414L3.414 10z\\" opacity=\\"0.5\\"></path>
                <path d=\\"M12.586 4l-1.293-1.293a1 1 0 111.414-1.414l3 3a1 1 0 010 1.414l-3 3a1 1 0 01-1.414-1.414L12.586 6H5a1 1 0 110-2h7.586z\\"></path>
              </g>
            </svg></div>
          <div class=\\"MuiListItemText-root-65\\"><span class=\\"MuiTypography-root-71 u-db u-ellipsis MuiTypography-subheading-78 MuiListItemText-primary-68\\"><div class=\\"TransactionLabel\\">Edf Particuliers</div><div><div class=\\"TransactionInfo\\"><span class=\\"TransactionInfoLabel\\">Date :</span>Saturday 26 August</div>
    </div></span>
  </div>
  </li>
  <li class=\\"MuiListItem-root-51 MuiListItem-default-54 MuiListItem-gutters-59\\">
    <div class=\\"MuiListItemIcon-root-64 WithStyles-ListItemIcon--root-63\\"></div>
    <div class=\\"MuiListItemText-root-65\\"><span class=\\"MuiTypography-root-71 u-db u-ellipsis MuiTypography-subheading-78 MuiListItemText-primary-68\\">Power</span></div><svg viewBox=\\"0 0 16 16\\" class=\\"styles__icon___23x3R\\" style=\\"fill: var(--coolGrey);\\" width=\\"16\\" height=\\"16\\">
      <path fill-rule=\\"evenodd\\" d=\\"M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414\\"></path>
    </svg>
  </li>
  <li class=\\"MuiListItem-root-51 MuiListItem-default-54 MuiListItem-gutters-59\\">
    <div class=\\"MuiListItemIcon-root-64 WithStyles-ListItemIcon--root-63\\"><svg width=\\"16\\" height=\\"16\\" class=\\"styles__icon___23x3R\\">
        <path d=\\"M13 1h3v2H0V1h3V0h2v1h6V0h2v1zM0 4h16v10.998A1 1 0 0115 16H1c-.552 0-1-.456-1-1.002V4zm2 2v8h12V6H2zm6 3h5v4H8V9z\\"></path>
      </svg></div>
    <div class=\\"MuiListItemText-root-65\\"><span class=\\"MuiTypography-root-71 u-db u-ellipsis MuiTypography-subheading-78 MuiListItemText-primary-68\\">Assigned to Aug 2017</span></div><svg viewBox=\\"0 0 16 16\\" class=\\"styles__icon___23x3R\\" style=\\"fill: var(--coolGrey);\\" width=\\"16\\" height=\\"16\\">
      <path fill-rule=\\"evenodd\\" d=\\"M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414\\"></path>
    </svg>
  </li>
  <li class=\\"MuiListItem-root-51 MuiListItem-default-54 MuiListItem-gutters-59\\">
    <div class=\\"MuiListItemIcon-root-64 WithStyles-ListItemIcon--root-63\\"><svg width=\\"16\\" height=\\"16\\" class=\\"styles__icon___23x3R\\">
        <path fill-rule=\\"evenodd\\" d=\\"M8 0C3.589 0 0 3.589 0 8s3.589 8 8 8a8.039 8.039 0 007.229-4.571 1 1 0 10-1.806-.858A6.031 6.031 0 018 14c-3.309 0-6-2.691-6-6s2.691-6 6-6c1.482 0 2.866.571 3.938 1.507a.989.989 0 00-.706.353 1 1 0 00.128 1.408l3 2.5A1 1 0 0016 7V3a1 1 0 00-1-1 .987.987 0 00-.96.801C12.548 1.08 10.364 0 8 0z\\"></path>
      </svg></div>
    <div class=\\"MuiListItemText-root-65\\"><span class=\\"MuiTypography-root-71 u-db u-ellipsis MuiTypography-subheading-78 MuiListItemText-primary-68\\"><div>Occasional transaction</div></span></div><svg viewBox=\\"0 0 16 16\\" class=\\"styles__icon___23x3R\\" style=\\"fill: var(--coolGrey);\\" width=\\"16\\" height=\\"16\\">
      <path fill-rule=\\"evenodd\\" d=\\"M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414\\"></path>
    </svg>
  </li>
  </ul>
</div>
</div>
</div>"
`;
